tests/test_validity.py

```diff
@@ -176,78 +209,92 @@ def action_reverse(cp_action):
 
         model_r += cp_action.n1 - cp_action.n0
         if cp_action.clear_adj_deps:
-            data.clear()
+            data.difference_update(range(cp_action.n0, cp_action.n1))
```

Comment:
Should this be `data.clear()`?

Reply:
`data` stores the forward data used in the adjoint computation. So, after using the forward data, one expected it would not be necessary to keep it stored anymore. That is the reason to use `data.clear`.

Reply:
If the reverse action clears only storage associated with the reversed steps then
```
Forward(0, 2, False, True, StorageType.WORK)
EndForward()
Reverse(2, 1, True)
Reverse(1, 0, True)
EndReverse()
```
would be valid. If this is _invalid_ (which I think it should be) then we should have `data.clear()` in the validity test.

Reply:
I've updated the test, assuming that `data.clear()` should be used here.